### ISSUE TITLE
BUG: Macro collision (`complex`) with Windows SDK in amos code

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#20291]
While building 1.13.0rc1 in [conda-forge](https://github.com/conda-forge/scipy-feedstock/pull/270), I'm running into macro collisions with the windows SDK:

```
[64/1477] Compiling C object scipy/special/lib_amos.a.p/_amos.c.obj
FAILED: scipy/special/lib_amos.a.p/_amos.c.obj 
"clang-cl" "-Iscipy\special\lib_amos.a.p" "-Iscipy\special" "-I..\scipy\special" "-Iscipy\_lib" "-I..\scipy\_lib" "-I..\scipy\_build_utils\src" "-DNDEBUG" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/clang:-std=c99" "/O2" "-Wno-unused-but-set-variable" "-Wno-unused-function" "-Wno-conversion" "-Wno-misleading-indentation" "-D_USE_MATH_DEFINES" "/Fdscipy\special\lib_amos.a.p\_amos.c.pdb" /Foscipy/special/lib_amos.a.p/_amos.c.obj "/c" ../scipy/special/_amos.c
In file included from ../scipy/special/_amos.c:96:
../scipy/special\_amos.h(106,15): error: expected ';' after top level declarator
  106 | double complex amos_airy(double complex, int, int, int *, int *);
      |               ^
      |               ;
../scipy/special\_amos.h(107,61): error: redefinition of parameter '_complex'
  107 | int amos_besh(double complex, double, int, int, int, double complex *, int *);
      |                                                             ^
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(44,29): note: expanded from macro 'complex'
   44 |             #define complex _complex
      |                             ^
```

I'm not sure if this would be solved by `WIN32_LEAN_AND_MEAN`, but in any case, looking at `corecrt_math.h` directly, the code shows that we could disable it by setting `_COMPLEX_DEFINED` (or investigating how `_CRT_INTERNAL_NONSTDC_NAMES` can be disabled).

```C
    // Definition of the _complex struct to be used by those who use the complex
    // functions and want type checking.
    #ifndef _COMPLEX_DEFINED
        #define _COMPLEX_DEFINED

        struct _complex
        {
            double x, y; // real and imaginary parts
        };

        #if defined(_CRT_INTERNAL_NONSTDC_NAMES) && _CRT_INTERNAL_NONSTDC_NAMES && !defined __cplusplus
            // Non-ANSI name for compatibility
            #define complex _complex
        #endif
    #endif
```

PS. This is ultimately a consequence of https://github.com/scipy/scipy/pull/19587, so CC @ilayn

**Comment by ilayn** [comment]
Since you have much more exposure to all kinds of systems, whatever makes your life easier I am fine with it. I just tried to make tests pass on CI and unfortunately do not have the best-practice information about `complex` definitions in the ecosystem. We can also steal the `complex` implementation mechanisms from the libraries we vendor.

**Comment by tylerjereddy** [comment]
@h-vetinari is this a blocker for RC2? It looks like it may be? Do you want to open a PR with the attempted fix?

With the new `pybind11` release now up on PyPI, NumPy `2.0.0rc1` may be imminent so I'd like us to be getting closer to stable release.

**Comment by h-vetinari** [comment]
Yeah, this is a blocker for conda-forge (which doesn't need to hold up rc2, but definitely GA) - we're simply unable to build with clang-cl for now it seems.

I've tried building with `-D_COMPLEX_DEFINED`, and it still errors:
```
[64/1477] Compiling C object scipy/special/lib_amos.a.p/_amos.c.obj
FAILED: scipy/special/lib_amos.a.p/_amos.c.obj 
"clang-cl" "-Iscipy\special\lib_amos.a.p" "-Iscipy\special" "-I..\scipy\special" "-Iscipy\_lib" "-I..\scipy\_lib" "-I..\scipy\_build_utils\src" "-DNDEBUG" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/clang:-std=c99" "/O2" "-Wno-unused-but-set-variable" "-Wno-unused-function" "-Wno-conversion" "-Wno-misleading-indentation" "-D_COMPLEX_DEFINED" "-D_USE_MATH_DEFINES" "/Fdscipy\special\lib_amos.a.p\_amos.c.pdb" /Foscipy/special/lib_amos.a.p/_amos.c.obj "/c" ../scipy/special/_amos.c
In file included from ../scipy/special/_amos.c:96:
../scipy/special\_amos.h(106,15): error: expected ';' after top level declarator
  106 | double complex amos_airy(double complex, int, int, int *, int *);
      |               ^
      |               ;
../scipy/special\_amos.h(107,61): error: redefinition of parameter 'complex'
  107 | int amos_besh(double complex, double, int, int, int, double complex *, int *);
      |                                                             ^
../scipy/special\_amos.h(107,22): note: previous declaration is here
  107 | int amos_besh(double complex, double, int, int, int, double complex *, int *);
      |                      ^
```

Unfortunately I'm without a laptop until the 7th of April. I can review & merge things on the phone, but not write patches myself.

**Comment by h-vetinari** [comment]
I can try one more thing in https://github.com/conda-forge/scipy-feedstock/pull/270, which is to set `-D_CRT_INTERNAL_NONSTDC_NAMES=0`, though even if this works, it's not really a releaseable solution IMO.

**Comment by ilayn** [comment]
I think this is not a macro collision but clang thinks `double complex` is a function called `complex` with output type `double`. Or something like that. 

I mean...

Anyways, if we can find what clang-cl thinks how `double complex` data type should be defined then we should be able to fix it probably.

**Comment by h-vetinari** [comment]
If you have conda installed, you can install the following dependencies (preferably into a fresh environment)
```
    - vs2019_win-64
    - clang
    - flang <18
    - lld
    - compiler-rt
    - libblas =*=*netlib
    - libcblas
    - liblapack
    - python
    - cython
    - meson-python
    - ninja
    - pkg-config
    - python-build
    - pybind11
    - pythran
    - numpy
    - pip
```
and then run our [`bld.bat`](https://github.com/conda-forge/scipy-feedstock/blob/main/recipe/bld.bat) from the SciPy root (minor adaptations would likely be necessary, e.g. replacing `%BUILD_PREFIX%` with `%CONDA_PREFIX%`).

Obviously it would also be possible to invoke conda-build on the recipe directly (using a checkout of my PR against the feedstock), but that's a bit harder to debug.

**Comment by h-vetinari** [comment]
> If you have conda installed, you can install the following dependencies (preferably into a fresh environment)

Ah, and because conda-forge cannot ship the actual msvc compiler, you also need to have VS2019 install (or at least the vc142 tool chain, if you're on VS2022 already)

**Comment by rgommers** [comment]
> Yeah, this is a blocker for conda-forge (which doesn't need to hold up rc2, but definitely GA) - we're simply unable to build with clang-cl for now it seems.

As a general comment: I don't think this is blocking for a final 1.13.0 release for more than a couple of days. It's quite urgent to do a release shortly after NumPy 2.0.0rc1. If we can't figure this one out _and_ the problematic PR cannot be easily reverted in the 1.13.x branch, then SciPy 1.13.0 package in conda-forge can just be missing for Windows only for a little while.

**Comment by ilayn** [comment]
For some reason, I can't make it work with clang-cl yet. So I think I will need some help with this one. 

Regretting touching C/C++ every time specifically because of silly stuff like this.

**Comment by ilayn** [comment]
OK apparently there is no way to make this code work with non-C99 compliant compiler (here is one such claim [1] but there are many others scattered around online). So either we convert this to C++ using `std::complex` everywhere or we go deep into rolling our own macros.

What do you think @steppi @izaid? 

[1] : https://stackoverflow.com/questions/1063406/c99-complex-support-with-visual-studio

**Comment by ilayn** [comment]
Then can I leave it to you folks? I'm a bit running out of steam with `special` code.

**Comment by steppi** [comment]
> Then can I leave it to you folks? I'm a bit running out of steam with `special` code.

Yes. I’ll do it today.

**Comment by izaid** [comment]
I ran into this recently as well with the conversion of specfun. Yeah, the issue is specifically `complex` is apparently non-compliant and clang doesn't like it. I was unaware of that before! The types that always seem to work are `float _Complex` or `double _Complex`, but even then I've found some of the C complex functions undefined.

To make things work, I put everything as `std::complex<double>` in the source files and, if an `extern "C"` wrapper was needed, casted to and from `double _Complex`.

**Comment by rgommers** [comment]
See this for the full story of using complex numbers in C/C++ in a portable way: https://github.com/lysnikolaou/c-cpp-complex. (Cc @lysnikolaou for visibility)

tl;dr it's very complex

**Comment by lysnikolaou** [comment]
FWIW if there's no downside in turning doing this in C++, I think it'll be much easier to maintain and definitely worth the effort.

**Comment by tylerjereddy** [comment]
> Yes. I’ll do it today.

@steppi if this proves harder/more problematic than you thought can you ping me so I can think of a plan for `1.13.0`? Thanks!

**Comment by steppi** [comment]
> > Yes. I’ll do it today.
> 
> @steppi if this proves harder/more problematic than you thought can you ping me so I can think of a plan for `1.13.0`? Thanks!

@izaid’s working on it now. I don’t anticipate it being too difficult, similar to specfun, which he just did recently, but fewer lines of code. Do you need this done today? If he doesn’t finish by tonight I’ll take over for him.

**Comment by tylerjereddy** [comment]
I doubt it needs to be done today, I still have to build new OpenBLAS binaries off their `develop` branch when Matti sends me instructions. Keeping me posted if it gets blocked by something unanticipated would be good though.

**Comment by steppi** [comment]
@tylerjereddy, I've reviewed and merged @izaid's fix. Things should be good now.
